WordSearchV2.0

EXR-38 opened the pull request:
Handle the exception when the user doesn't
input any argument.
Check whether the file is empty before proceeding.
Remove unused methods.

EXR-38 committed:
WordSearchV2.0

Handle the exception when the user doesn't
input any argument.
Check whether the file is empty before proceeding.
Remove unused methods.

premsrj commented:
@Abhijith-Sajimon Why create a new PR instead of updating the old one itself?

premsrj commented:
Please add me and @kssreejith1994  to reviewers

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+
+    public static void main(String[] args){
+
+        if(args.length!=2)
```
```suggestion
        if(args.length!=2) {

```

Opening bracket should be in the same line

EXR-38 replied:
Sure. I will keep that in mind.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+        if(args.length!=2)
+        {
+            System.out.println("File path and search word cannot be empty");
+            System.exit(0);
```
Is there a reason exit() is used instead of a return statement?

EXR-38 replied:
No. Will change that.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+            System.out.println("File path and search word cannot be empty");
+            System.exit(0);
+        }
+        String path=args[0];
```
There should be a space before and after =
Here and everywhere else in the code

EXR-38 replied:
Sure. Will do.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+        String search=args[1];
+        File file=new File(path);
+        System.out.println("Processing...");
+        doesFileExist(file);
```
By this logic, the wordCount method will be called even if the file does not exist. You should either throw an exception from doesFileExist or 
Make it return a boolean, check the return value and only then call wordCount

EXR-38 replied:
I have made doesFileExist() return a boolean value. wordCount() will be called only if doesFileExist() true.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+        if (!file.isFile())
+        {
+            System.out.println("File does not exist");
+        } else isFileFormatSupported(file);
```
doesFileExist() method should only be concerned with the existence of the file. It should not be calling isFileFormatSupported() from within. Either this second method should be called externally, or you have to rename doesFileExist to validateFile() or something with a similar meaning

kssreejith1994 replied:
Also, it would be better to use braces even if its a single line statement.

EXR-38 replied:
Sure. I will keep that in mind.

EXR-38 replied:
@premsrj I have  changed it to validateFile() and made all the changes per you suggested.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+            e.printStackTrace();
+        }
+        if(data!=null) {
+                removeSpecialChar = data.replaceAll("\\p{Punct}", "");
```
Indentation is wrong

EXR-38 replied:
I have corrected the indentation.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+            e.printStackTrace();
+        }
+        if(data!=null) {
+                removeSpecialChar = data.replaceAll("\\p{Punct}", "");
```
Additionally, you really don't need two strings here 'data' and 'removeSpecialChar'
This could just be
data = data.replaceAll(...)

EXR-38 replied:
Sure will do

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+        }
+        if(data!=null) {
+                removeSpecialChar = data.replaceAll("\\p{Punct}", "");
+        }else {
```
```suggestion
        } else {

```
Space before else

EXR-38 replied:
Sure. Will do.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+                removeSpecialChar = data.replaceAll("\\p{Punct}", "");
+        }else {
+                System.out.println("Empty file");
+                System.exit(1);
```
What is the reason for using explicit exit() call instead of a return statement?

EXR-38 replied:
I will use return statement instead of explicit exit().

premsrj commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+            System.out.println("Word found");
+            System.out.println(search + " occurs " + wordcount + " times in the file");
+        }
+        return wordcount;
```
If this wordCount is not used in the main method, is it required to be returned? Better to have this as a void method, right?

EXR-38 replied:
Yes. I will change it to void.

kssreejith1994 commented on Main.java:
```diff
@@ -0,0 +1,73 @@
+        int wordcount = 0;
+        String temp = "";
+        while (st.hasMoreTokens()) {
+            temp = st.nextToken();
```
You can directly use st.nextToken inside the equalsIgnoreCase. No need to declare another variable since its not used else where.

EXR-38 replied:
Sure. Will do.

Abhijith-Sajimon commented:
> @Abhijith-Sajimon Why create a new PR instead of updating the old one itself?

There was no need. I did that by mistake.

Abhijith-Sajimon commented:
> Please add me and @kssreejith1994 to reviewers

Sure Prem.

EXR-38 committed:
WordSearchV3.0

Used return statement instead of explicit exit() when possible.
Removed unneeded variables.
Corrected the placement of curly braces and added spaces wherever needed.
Corrected the indentation.

kssreejith1994 commented on Main.java:
```diff
@@ -5,61 +5,58 @@
-        String search=args[1];
-        File file=new File(path);
+        String path = args[0];
+        String search = args[1];
```
Can you rename this to a more understandable variable name?

EXR-38 replied:
Are you referring to path and search?

EXR-38 replied:
I have renamed them to textFilePath and searchKeyword respectively.

kssreejith1994 commented on Main.java:
```diff
@@ -5,61 +5,58 @@
+            System.exit(0);
+        } else {
+            String fullName = file.getName();
+            if (!(fullName.endsWith(".txt") || fullName.endsWith(".json"))) {
```
What if used passes an extension other than .txt and .json?

EXR-38 replied:
It will show "File format not supported"

kssreejith1994 replied:
So it does not need the explicit check for .json right?

EXR-38 replied:
The code needs to support both .txt and .json

kssreejith1994 commented on Main.java:
```diff
@@ -5,61 +5,58 @@
-                System.exit(1);
-            }
-        StringTokenizer st = new StringTokenizer(removeSpecialChar);
+        if (data != null) {
```
Data will never be null since data you are setting as data="" in line 41

EXR-38 replied:
But when I passed an empty file, the code worked fine and it showed "File does not contain any data". So the else condition works, right?

EXR-38 committed:
WordSearchV4.0

Renamed variables "path" and "search" to "textFilePath" and "searchKeyword".

premsrj commented on Main.java:
```diff
@@ -0,0 +1,66 @@
+        String searchKeyword = args[1];
+        File file = new File(textFilePath);
+        System.out.println("Processing...");
+        validateFile(file);
```
Why is this method called here?

EXR-38 replied:
It was there by mistake. I have removed it.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,66 @@
+            System.out.println("File format not supported");
+            System.exit(0);
+        }
+        return true;
```
Why is a true being returned here? It does not seem to have anything to do with whether the file is valid or not

EXR-38 replied:
I have changed the method to return true when the file type is either a .txt or .json file and false when it is of any other type.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,66 @@
+        if (!file.exists()) {
+            System.out.println("File does not exist");
+
+        } else if (!(file.getName().endsWith(".txt") || file.getName().endsWith(".json"))) {
```
Use static final constants instead of hardcoding .txt and .json

EXR-38 replied:
Sure. Will do that.

premsrj commented:
@Abhijith-Sajimon Please re-request for review from me and @kssreejith1994 when you are done making changes.

premsrj commented on Main.java:
```diff
@@ -0,0 +1,66 @@
+
+        } else if (!(file.getName().endsWith(".txt") || file.getName().endsWith(".json"))) {
+            System.out.println("File format not supported");
+            System.exit(0);
```
Do not use System.exit(), just return a boolean value to indicate whether the file is valid or not, and in the main method check for the return value and proceed accordingly.

EXR-38 replied:
Sure. I will do that.

premsrj replied:
Still using exit() method here. Just return false here

EXR-38 replied:
I have removed the explicit exit() statement and used return instead.

EXR-38 committed:
WordSearchV5.0

Created static constants for ".txt", ".json and the "\\p{punct}".
Removed unnecessary method calls.

EXR-38 committed:
WordSearchV6.0

Add feature to read the file with multiple lines of data

premsrj commented on Main.java:
```diff
@@ -0,0 +1,75 @@
+            if (fileData == null) {
+                System.out.println("File does not contain any data");
+                return;
+            } else
```
use brackets for else

```suggestion
            } else {

```

EXR-38 replied:
Sure. Will do that.

kssreejith1994 commented on Main.java:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    private static boolean validateFile(File file) {
```
Change this to isValidFile

EXR-38 replied:
Sure. Will do that.

kssreejith1994 commented on Main.java:
```diff
@@ -0,0 +1,75 @@
+
+        if (!file.exists()) {
+            System.out.println("File does not exist");
+            System.exit(0);
```
Use return instead of exit or return false from here itself.

EXR-38 replied:
Sure. Will do that.

EXR-38 committed:
WordSearchV7.0

Removed System.exit() and used return instead.
Added curly brackets for else statement.

EXR-38 committed:
WordSearchV8.0

Changed method name from validateFile() to isValidFile()
